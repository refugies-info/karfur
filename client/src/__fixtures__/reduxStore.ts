import { RootState } from "services/rootReducer";
import { SearchResultsState } from "services/SearchResults/searchResults.reducer";
import { UserFavoritesState } from "services/UserFavoritesInLocale/UserFavoritesInLocale.reducer";
import { activeThemesMock } from "./activeThemes";

const initialMockUserState = {
  user: null,
  admin: false,
  traducteur: false,
  expertTrad: false,
  contributeur: false,
  membreStruct: false,
  userId: "",
  userFetched: false,
  rolesInStructure: [],
};
const initialMockLangueState = {
  langues: [],
  languei18nCode: "fr",
  showLanguageModal: false,
  showLangModal: false,
};

const initialMockTranslationState = {
  translation: {
    initialText: {},
    translatedText: {},
  },
  translations: [],
};

const initialUserFavoritesState: UserFavoritesState = {
  favorites: [],
  showFavoriteModal: false
}

const initialMockSearchReults: SearchResultsState = {
  results: {
    dispositifs: [],
    demarches: [],
    dispositifsSecondaryTheme: []
  },
  query: {
    search: "",
    departments: [],
    themes: [],
    needs: [],
    age: [],
    frenchLevel: [],
    language: [],
    sort: "date",
    type: "all",
  }
}

export const initialMockStore = {
  user: initialMockUserState,
  langue: initialMockLangueState,
  activeDispositifs: [],
  tts: { ttsActive: false, showAudioSpinner: false },
  userStructure: null,
  selectedDispositif: null,
  loadingStatus: {},
  translation: initialMockTranslationState,
  activeStructures: [],
  selectedStructure: null,
  allDispositifs: [],
  allStructures: [],
  users: [],
  userFavorites: initialUserFavoritesState,
  userContributions: [],
  needs: [],
  themes: {
    activeThemes: activeThemesMock,
    inactiveThemes: []
  },
  searchResults: initialMockSearchReults
};
