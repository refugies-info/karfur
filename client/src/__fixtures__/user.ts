import { GetUserInfoResponse, UserStatus } from "api-types";
import moment from "moment";

const date = moment("2019-05-01 12:00:00");
export const testUser: GetUserInfoResponse = {
  username: "test",
  _id: "testObjectId",
  structures: [],
  contributions: [],
  phone: "",
  roles: [],
  email: "",
  selectedLanguages: [],
  status: UserStatus.ACTIVE,
  traductionsFaites: []
};

export const testUserWithRoles: GetUserInfoResponse = {
  ...testUser,
  structures: ["testObjectId"],
  roles: [
    { nom: "Admin", _id: "testObjectId", nomPublic: "Admin" },
    { nom: "Trad", _id: "testObjectId", nomPublic: "Trad" },
    {
      nom: "ExpertTrad",
      _id: "testObjectId",
      nomPublic: "ExpertTrad"
    },
    {
      nom: "Contrib",
      _id: "testObjectId",
      nomPublic: "Contrib"
    }
  ]
};
