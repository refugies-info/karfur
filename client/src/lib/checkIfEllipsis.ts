export const checkIfEllipsis = (element: HTMLElement) => element.offsetWidth < element.scrollWidth;
