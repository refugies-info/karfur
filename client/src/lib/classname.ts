export const cls = (...classNames: (string | boolean)[]) => classNames.filter(className => !!className).join(" ");
