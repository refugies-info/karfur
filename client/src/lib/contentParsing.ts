export const getCalloutTranslationKey = (level: "info" | "important") => {
  switch (level) {
    case "info":
      return "callout_info"
    case "important":
      return "callout_important"
  }
}

interface ToParse {
  nodeAttr: string;
  translation: string;
}

export const translationParsing = (originalHTML: string, toParse: ToParse[]) => {
  let parsedHTML = originalHTML;
  for (const parsing of toParse) {
    parsedHTML = parsedHTML.replace(parsing.nodeAttr, `${parsing.nodeAttr} data-title='${parsing.translation}'`)
  }
  return parsedHTML
}
