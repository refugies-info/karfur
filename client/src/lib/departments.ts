import { departments } from "data/departments"

export const getDbDepartment = (depName: string) => {
  const index = departments.map(dep => dep.split(" - ")[1]).indexOf(depName);
  return index > -1 ? departments[index] : depName;
}

/**
 * Get department name or postal code from GMap autocomplete
 * @returns a department number or name formatted
 */
export const formatDepartment = (depName: string) => {
  let depDbFormat = depName;

  // postal code
  if (depName.match(/\d{5}/g)) return depName.slice(0, 2);

  // department
  if (!depDbFormat.includes(" - ")) {
    depDbFormat = getDbDepartment(depName);
  }

  const dep = depDbFormat.split(" - ");
  if (!dep[1]) return dep[0];
  return `${dep[1]} (${dep[0]})`;
}

export const getDepartmentFromNumber = (number: string): string => {
  return departments.find(d => d.startsWith(number)) || "";
}
