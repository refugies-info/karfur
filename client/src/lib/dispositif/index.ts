export * from "./isStatus";
export * from "./getMetadataText";
export * from "./canEdit";
export * from "./canTranslate";
export * from "./getTranslationPageData";
