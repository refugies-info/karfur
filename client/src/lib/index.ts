export const jsUcfirst = (string: string) =>
  string &&
  string.length > 1 &&
  string.charAt(0).toUpperCase() + string.slice(1, string.length);

export const jsLcfirst = (string: string) =>
  string &&
  string.length > 1 &&
  string.charAt(0).toLowerCase() + string.slice(1, string.length);

export const limitNbCaracters = (string: string, nbCaractersMax: number) =>
  string.substring(0, Math.min(string.length, nbCaractersMax)) +
  (string.length > nbCaractersMax ? "..." : "");

export const removeAccents = (str = "") => {
  var accents =
    "ÀÁÂÃÄÅàáâãäåÒÓÔÕÕÖØòóôõöøÈÉÊËèéêëðÇçÐÌÍÎÏìíîïÙÚÛÜùúûüÑñŠšŸÿýŽž";
  var accentsOut =
    "AAAAAAaaaaaaOOOOOOOooooooEEEEeeeeeCcDIIIIiiiiUUUUuuuuNnSsYyyZz";
  const chars = str.split("");
  let i, x;
  for (i = 0; i < chars.length; i++) {
    if ((x = accents.indexOf(chars[i])) !== -1) {
      chars[i] = accentsOut[x];
    }
  }
  return chars.join("");
};

// rules for password check :
// nbNumUpper = nbNumeric + nbUpper
// score 0 (cannot set password): length < 5
// score 0.5 : 7>length>=5 + nbNumUpper = 1 or 2 or0
// score 1.5 : 7>length >= 5 + nbNumUpper>=3 OR length>=7 + nbNumUpper<3
// score 2.5 : length >= 7 + nbNumUpper>=3
// score 3.5 : length >= 7 + nbNumUpper>=3 + at least 1 special caractere

// make sure to use same function in backend
export const computePasswordStrengthScore = (
  password: string
): { score: number } => {
  if (password.length < 5) {
    return { score: 0 };
  }

  const numerics = password.match(/\d+/g)?.join("");
  const nbNumeric = numerics ? numerics.length : 0;

  const upperCase = password.match(/[A-Z]/g)?.join("");
  const nbUpperCase = upperCase ? upperCase.length : 0;

  if (password.length < 7) {
    if (nbUpperCase + nbNumeric < 3) {
      return { score: 1 };
    }
    return { score: 2 };
  }

  if (nbNumeric + nbUpperCase < 3) {
    return { score: 2 };
  }
  const specialCar = password.match(/[!@#$&*]/g)?.join("");
  const nbSpecialCar = specialCar ? specialCar.length : 0;
  if (nbSpecialCar > 0) {
    return { score: 4 };
  }

  return { score: 3 };
};
