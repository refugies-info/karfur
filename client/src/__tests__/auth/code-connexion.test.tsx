import { wrapWithProvidersAndRender } from "../../../jest/lib/wrapWithProvidersAndRender";
import codeConnexion from "../../pages/auth/code-connexion";
import mockRouter from "next-router-mock";
import { initialMockStore } from "__fixtures__/reduxStore";
import { act, ReactTestRenderer } from "react-test-renderer";
import { setupGoogleMock } from "../../__mocks__/react-google-autocomplete";
import "jest-styled-components";

jest.mock("next/router", () => require("next-router-mock"));

describe("auth/connexion", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    setupGoogleMock();
  });

  let component: ReactTestRenderer;

  it("renders null if no email", () => {
    act(() => {
      component = wrapWithProvidersAndRender({
        Component: codeConnexion,
        reduxState: {
          ...initialMockStore,
        },
      });
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
  it("renders code-connexion if email", () => {
    const url = "http://refugies.info/auth/code-connexion?email=test@example.com";
    Object.defineProperty(window, "location", {
      value: new URL(url),
    });
    act(() => {
      mockRouter.push("/auth/code-connexion?email=test@example.com");
    });
    act(() => {
      component = wrapWithProvidersAndRender({
        Component: codeConnexion,
        reduxState: {
          ...initialMockStore,
        },
      });
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
});
