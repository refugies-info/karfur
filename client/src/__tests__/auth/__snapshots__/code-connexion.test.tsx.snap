// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth/connexion renders code-connexion if email 1`] = `
<div
  className="container half"
>
  <button
    className="fr-btn fr-btn--tertiary fr-btn--sm fr-icon-arrow-left-line fr-btn--icon-left"
    id="fr-button-:r6:"
    onClick={[Function]}
    style={{}}
  >
    Retour
  </button>
  <div
    className="title"
  >
    <h1>
      Entrez le code reçu
    </h1>
    <p
      className="subtitle"
    >
      Un code temporaire à 6 chiffres vous a été envoyé à 
      test@example.com
      .
       
      <a
        className="text-decoration-underline"
        href="/auth"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Ce n'est pas vous ?
      </a>
    </p>
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="fr-input-group mb-0"
    >
      <label
        className="fr-label"
        htmlFor="input-:r7:"
      >
        Code de connexion temporaire
      </label>
      <input
        aria-describedby="input-:r7:-desc-error"
        autoFocus={true}
        className="fr-input"
        id="input-:r7:"
        name="code"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <button
      className="fr-btn fr-icon-check-line fr-btn--icon-right button mt mb-4"
      id="fr-button-:r8:"
      style={{}}
      type="submit"
    >
      Valider
    </button>
    <button
      className="fr-btn fr-btn--tertiary fr-icon-mail-line fr-btn--icon-right button"
      id="fr-button-:r9:"
      onClick={[Function]}
      style={{}}
    >
      Renvoyer le code
    </button>
  </form>
  <div
    className="mb-4 space_top row"
  >
    <div
      className="col"
    >
      <a
        className="fr-btn fr-btn--tertiary button"
        href="https://mail.google.com/mail/u/0/"
        id="fr-button-:ra:"
        style={{}}
      >
        <img
          alt=""
          className="me-2"
          data-nimg="1"
          decoding="async"
          height={24}
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
          srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
          style={
            {
              "color": "transparent",
            }
          }
          width={24}
        />
        Ouvrir Gmail
      </a>
    </div>
    <div
      className="col"
    >
      <a
        className="fr-btn fr-btn--tertiary button"
        href="https://outlook.live.com/mail/0/"
        id="fr-button-:rb:"
        style={{}}
      >
        <img
          alt=""
          className="me-2"
          data-nimg="1"
          decoding="async"
          height={24}
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
          srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
          style={
            {
              "color": "transparent",
            }
          }
          width={24}
        />
        Ouvrir Outlook
      </a>
    </div>
  </div>
  <p
    className="small mt-6 text-center"
  >
    Pensez à vérifiez votre courrier indésirable !
  </p>
</div>
`;

exports[`auth/connexion renders null if no email 1`] = `null`;
